html/browsers/browsing-the-web/back-forward-cache/resources/executor.html

```diff
@@ -0,0 +1,117 @@
+<!DOCTYPE HTML>
+<script src="/resources/testdriver.js"></script>
+<script src="/resources/testdriver-vendor.js"></script>
+<script src="/resources/channel.sub.js"></script>
+<script>
+const params = new URLSearchParams(window.location.search)
+const uuid = params.get("uuid");
+let ctxChannel = null; // This is created in the initial load
+
+// --------
+// Recording events
+
+// The recorded events are stored in localStorage rather than global variables
+// to catch events fired just before navigating out.
+function getPushedItems(key) {
+  return JSON.parse(localStorage.getItem(key) || '[]');
+}
+
+function pushItem(key, value) {
+  const array = getPushedItems(key);
+  array.push(value);
+  localStorage.setItem(key, JSON.stringify(array));
+}
+
+function recordEvent(eventName) {
+  pushItem(uuid + '.observedEvents', eventName);
+}
+
+function getRecordedEvents() {
+  return getPushedItems(uuid + '.observedEvents');
+}
+
+// Records events fired on `window` and `document`, with names listed in
+// `eventNames`.
+function startRecordingEvents(eventNames) {
+  for (const eventName of eventNames) {
+    window.addEventListener(eventName, event => {
+      let result = eventName;
+      if (event.persisted) {
+        result += '.persisted';
+      }
+      if (eventName === 'visibilitychange') {
+        result += '.' + document.visibilityState;
+      }
+      recordEvent('window.' + result);
+    });
+    document.addEventListener(eventName, () => {
+      let result = eventName;
+      if (eventName === 'visibilitychange') {
+        result += '.' + document.visibilityState;
+      }
+      recordEvent('document.' + result);
+    });
+  }
+}
+
+// When a comma-separated list of event names are given as the `events`
+// parameter in the URL, start record the events of the given names.
+if (params.get('events')) {
+  startRecordingEvents(params.get('events').split(','));
+}
+
+// --------
+// Executor and BFCache detection
+
+// bfcache is typically disabled if there's an open network connection e.g.
+// a websocket. Therefore some care is needed to avoid websockets being
+// left open when navigating.
+
+// To navigate the page, postMessage on a remote context a message of the form
+// {name: "navigateTo", url: `${url}`}
+// Using postMessage here ensures we don't create an unnecessary
+// channel to send a result.
+
+// Then the calling page must call TestDriverRemote.pause() before continuing.
+// This ensures that no subsequent commands will be buffered in the underlying
+// websocket before the navigation starts. This has to be done in the caller
+// to avoid a race condition between the caller sending a new command and the
+// socket being closed.
```

Comment:
In https://github.com/web-platform-tests/wpt/pull/28950,

- I have something similar to `closeAllChannelSockets()` that suspends polling network requests on the remote context side,
- but no `pause()`-ish things, relying on that the caller doesn't send any new commands until the page becomes inactive

so I'm curious how `pause()` works.
Particularly, are both of `pause()` and `closeAllChannelSockets()` needed?
Is there a race condition between these two?